### ISSUE TITLE
config.py with set_LBFGS_options isn't included into installation?

### ISSUE THREAD

**Issue by sashahexe** [lululxvi/deepxde#364]
Dear Lu Lu, 

Could you please help with usage of L-BFGS-B. I am trying to get some results with it but run into the problem that number of iterations exceeds the limit. I tried to find a backend file where one could tweak the parameters. Here on the github I can see you have a file config.py in the optimizers folder (
https://github.com/lululxvi/deepxde/blob/d5e9466c6f542bd50d3c6b901be40d6948ed4c5f/deepxde/optimizers/config.py), but in my installations in Ubuntu and in Colab this file is missing. Could you please give a hint where to find these 

`def set_LBFGS_options(
    maxcor=100,
    ftol=0,
    gtol=1e-8,
    maxiter=15000,
    maxfun=None,
    maxls=50,
):`

Second question please: Is it possible to set the maximum training/testing losses for LBFGS to stop and not only the difference between test and train?

**Comment by lululxvi** [comment]
-This file is added very recently and is not in the public version yet. You have to download the source code directly from Github.
- You can set `ftol` so that L-BFGS will stop once the training loss is smaller than `ftol`.

**Comment by sashahexe** [comment]
Thank you for the speedy answer!

I have installed the deepxde from github straight into the colab to test these options out. 

It seems I am doing something wrong when tweaking the optimizer's parameters in the congig.py. For example I have set the ftol to 1.25e-4 but the process has terminated with the losses:

```
INFO:tensorflow:Optimization terminated with:
  Message: b'CONVERGENCE: REL_REDUCTION_OF_F_<=_FACTR*EPSMCH'
  Objective function value: 0.001751
  Number of iterations: 9300
  Number of functions evaluations: 21656

```
```
Best model at step 21656:
  train loss: 1.75e-03
  test loss: 1.46e-03
  test metric: []
```


I have also set the 
`
maxiter=250000
`
but if it doesn't stop because of REL_REDUCTION_OF_F_<=_FACTR*EPSMCH' then it stops as 15000 steps point. 

Here is my colab 

https://colab.research.google.com/drive/12-NA5gNnRRqN863j17K2pGRlnGT8QoP2?usp=sharing

If in the submenu Runtime you choose to "run all" it will automatically install the deepxde from the gitgub and make the changes to the config.py in the optimizers folder, afterwards it will start the PINN training. 

Could you please have a look, I can't find what is wrong there.

**Comment by sashahexe** [comment]
Thank you for the speedy answer!

I have installed the deepxde from github straight into the colab to test these options out. 

It seems I am doing something wrong when tweaking the optimizer's parameters in the config.py. For example I have set the `ftol` to 1.25e-4 but the process has terminated with the losses:

```
INFO:tensorflow:Optimization terminated with:
  Message: b'CONVERGENCE: REL_REDUCTION_OF_F_<=_FACTR*EPSMCH'
  Objective function value: 0.001751
  Number of iterations: 9300
  Number of functions evaluations: 21656

```
```
Best model at step 21656:
  train loss: 1.75e-03
  test loss: 1.46e-03
  test metric: []
```


I have also set the 
`
maxiter=250000
`
but if it doesn't stop because of REL_REDUCTION_OF_F_<=_FACTR*EPSMCH' then it stops as 15000 steps point. 

Here is my colab 

https://colab.research.google.com/drive/1lLs58XWOSiSkhKBYttEBOPgO5ipiRlqY?usp=sharing

If in the submenu Runtime you choose to "run all" it will automatically install the deepxde from the github and make the changes to the config.py in the optimizers folder, afterwards it will start the PINN training.

**Comment by sashahexe** [comment]
I have found a useful thread about the part of my question regarding the REL_REDUCTION_OF_F_<=_FACTR*EPSMCH':

https://stackoverflow.com/questions/45616929/scipy-optimize-minimize-always-exit-with-rel-reduction-of-f

So apparently the precision of the machine while calculations (the EPSMCH) play a big role. I changed to float64 via `dde.config.real.set_float64()` which I found in issue 247 (https://github.com/lululxvi/deepxde/issues/247). That did the trick so the L-BFGS-B doesn't stop too early. 

I also managed to overcome the issue of `maxiter=15000` by firstly making a my own repository in github with a copy of deepxde where I pre-edited config.py to make sure colab does `import deepxde` with the right parameters for L-BFGS-B. 

**So at the moment I don't have technical issues left.**


The only thing I'm not sure about is this `ftol`.

They say:

`ftol = factr * numpy.finfo(float).eps.`

I.e., `factr `multiplies the default machine floating-point precision to arrive at `ftol` and 
This algorithm stops either if (1) ' (f^k - f^{k+1})/max{|f^k|,|f^{k+1}|,1} <= factr * eps ' - so ftol is the normalized difference between two consequent iterations and not the loss in terms of what PINN does, isn't it?

**Comment by lululxvi** [comment]
- It is always a good strategy to use float64 with L-BFGS.
- `ftol` is very confusing. I also checked online, but still are not very sure what is going on axactly.
- it is strange that you have to make a copy. There could be something related to colab.

**Comment by andrea-sacchetti** [comment]
Personally, I have successfully tweaked the "L-BFGS-B" options by means of invoking:

```
model.train_step.optimizer_kwargs = {'options': {'maxfun': 1e5, 'ftol': 1e-20, 'gtol': 1e-20, 'eps': 1e-20, 'iprint': -1, 'maxiter': 1e5}}
```

before 

```
model.train()
```

That seems to do the trick. It possibly only works with the "tensorflow.compat.v1" backend.

**Comment by lululxvi** [comment]
@andrea-sacchetti Differnet backends uses different L-BFGS solvers. "tensorflow.compat.v1" backend uses L-BFGS in Scipy, so your approach is totally fine. `set_LBFGS_options()` works for all backends, and has exactly the same effect as your approach for "tensorflow.compat.v1".
